src/sentry/api/endpoints/user_notification_providers.py

```diff
@@ -0,0 +1,84 @@
+from django.db import router, transaction
+from rest_framework import status
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry.api.base import control_silo_endpoint
+from sentry.api.bases.user import UserEndpoint
+from sentry.api.exceptions import ParameterValidationError
+from sentry.api.serializers import serialize
+from sentry.api.validators.notifications import validate_type
+from sentry.models import User
+from sentry.models.notificationsettingprovider import NotificationSettingProvider
+from sentry.notifications.serializers import NotificationSettingsProviderSerializer
+from sentry.notifications.types import NotificationSettingsOptionEnum
+from sentry.notifications.validators import (
+    UserNotificationSettingsProvidersDetailsSerializer,
+    allowed_providers,
+)
+
+
+@control_silo_endpoint
+class UserNotificationProvidersEndpoint(UserEndpoint):
+    # TODO(Steve): Make not private when we launch new system
+    private = True
+
+    def get(self, request: Request, user: User) -> Response:
+        """
+        Retrieve the notification provider preferences for a user.
+        Returns a list of NotificationSettingProvider rows.
+        """
+        notifications_settings = NotificationSettingProvider.objects.filter(
+            user_id=user.id,
+        )
+        notification_type = request.GET.get("type")
+        if notification_type:
+            try:
+                validate_type(notification_type)
+            except ParameterValidationError:
+                return self.respond({"type": ["Invalid type"]}, status=400)
+            notifications_settings = notifications_settings.filter(
+                type=notification_type,
+            )
+
+        notification_preferences = serialize(
+            list(notifications_settings), request.user, NotificationSettingsProviderSerializer()
+        )
+
+        return Response(notification_preferences)
+
+    def put(self, request: Request, user: User) -> Response:
```

Comment:
if i'm reading this correctly, this will update rows for providers that were not included in the request body? is this intentional?

Reply:
@JoshFerge yea, it's a limitation on how the form field works where it sends a list of providers. To do it the "right" way requires a lot more hacking in the UI.